File: PurpleUI-iOS/Sources/PurpleUI/Tokens/PUShadows.swift
import SwiftUI

// ----------------------------------------------------------------
// SHADOW TOKENS
//
// `tokens.json` (`shadow.*`) and `specs/foundation-shadows.json` ship these
// as CSS box-shadow strings: `offsetX offsetY blurRadius spreadRadius color`.
// SwiftUI's `.shadow(color:radius:x:y:)` has no direct box-shadow equivalent
// and no spread parameter, so each token below is converted from its web
// source value the same way `purpleui-react-native/src/tokens/shadows.ts`
// does, so all three platforms agree:
//
//   x (offset)   = CSS offsetX, unchanged
//   y (offset)   = CSS offsetY, unchanged
//   radius       = CSS blurRadius / 2  (CSS blur is a diameter-ish spread;
//                  SwiftUI's `radius` maps roughly 1:2 against it — this is
//                  the same halving RN's `shadowRadius: blurRadius / 2` uses)
//   color/opacity = taken straight from the rgba() alpha component
//   spread        = dropped (SwiftUI has no spread parameter — RN drops it
//                  too, since `shadowRadius`/`elevation` have no spread
//                  concept either)
//
// See each entry below for the original CSS box-shadow string it was
// converted from.
// ----------------------------------------------------------------

/// Named shadow tokens matching the Purple UI shadow scale.
///
/// Usage:
/// ```swift
/// myView
///     .shadow(color: PUShadows.bottomA.color, radius: PUShadows.bottomA.radius,
///             x: PUShadows.bottomA.x, y: PUShadows.bottomA.y)
/// ```
public struct PUShadow {
    public let color: Color
    public let radius: CGFloat
    public let x: CGFloat
    public let y: CGFloat

    // Pre-existing gap, fixed while in this file for another review item:
    // `public` stored properties alone only get an `internal` synthesized
    // memberwise initialiser, so consumers outside this module couldn't
    // construct their own `PUShadow` (e.g. for a one-off shadow that isn't
    // one of the named tokens below). Explicit `public init` added.
    public init(color: Color, radius: CGFloat, x: CGFloat, y: CGFloat) {
        self.color  = color
        self.radius = radius
        self.x      = x
        self.y      = y
    }
}

public struct PUShadows {
    /// Subtle card lift, button default — '0px 2px 2px 0px rgba(0,0,0,0.03)'
    public static let bottomA = PUShadow(color: .black.opacity(0.03), radius: 1,   x: 0, y: 2)
    /// Light card — '0px 2px 2px 0px rgba(0,0,0,0.06)'
    public static let bottomB = PUShadow(color: .black.opacity(0.06), radius: 1,   x: 0, y: 2)
    /// Floating elements, light — '0px 2px 10px 0px rgba(0,0,0,0.05)'
    public static let bottomC = PUShadow(color: .black.opacity(0.05), radius: 5,   x: 0, y: 2)
    /// Floating buttons, modals, heavy — '0px 2px 15px 0px rgba(0,0,0,0.10)'
    public static let bottomD = PUShadow(color: .black.opacity(0.10), radius: 7.5, x: 0, y: 2)

    /// Bottom sheet top shadow — '0px -2px 2px 0px rgba(0,0,0,0.03)'
    public static let topA = PUShadow(color: .black.opacity(0.03), radius: 1,   x: 0, y: -2)
    /// Sticky nav top shadow — '0px -2px 2px 0px rgba(0,0,0,0.06)'
    public static let topB = PUShadow(color: .black.opacity(0.06), radius: 1,   x: 0, y: -2)
    /// Bottom tray top edge — '0px -2px 2px 0px rgba(0,0,0,0.10)'
    public static let topC = PUShadow(color: .black.opacity(0.10), radius: 1,   x: 0, y: -2)
    /// Heavier top shadow — '0px -2px 15px 0px rgba(0,0,0,0.10)'
    public static let topD = PUShadow(color: .black.opacity(0.10), radius: 7.5, x: 0, y: -2)

    /// Side navigation drawer — '4px 0px 15px 0px rgba(0,0,0,0.20)'
    public static let sideNav = PUShadow(color: .black.opacity(0.20), radius: 7.5, x: 4, y: 0)

    /// Selected map tile — '0px 10px 15px -3px rgba(0,0,0,0.10)' (spread dropped)
    public static let tilesActive = PUShadow(color: .black.opacity(0.10), radius: 7.5, x: 0, y: 10)
    /// Hovered map tile, brand-tinted — '0px 10px 15px -3px rgba(116,88,253,0.40)'
    /// (spread dropped; rgba(116,88,253) === PUColors.brand)
    public static let tilesHover = PUShadow(color: PUColors.brand.opacity(0.40), radius: 7.5, x: 0, y: 10)
}

File: PurpleUI-Android/purpleui/src/main/kotlin/ai/purple/purpleui/tokens/PUShadows.kt
package ai.purple.purpleui.tokens

import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.unit.dp

/**
 * Shadow helpers matching the Purple UI named shadow tokens.
 *
 * `tokens.json` ships these as CSS box-shadow strings
 * (`offsetX offsetY blurRadius spreadRadius color`). Compose's `Modifier.shadow`
 * is a genuinely different rendering model - it takes a single non-directional
 * `elevation` plus `ambientColor`/`spotColor`, not an offset/blur/spread quad -
 * so there is no exact 1:1 conversion.
 *
 * Each token below is approximated from its web source value (see the comment
 * on each entry for the original CSS string), the same way
 * `purpleui-react-native/src/tokens/shadows.ts` documents its RN conversion.
 * The `elevation` chosen for each token mirrors RN's own `elevation` field for
 * the same named shadow, so the visual weight stays roughly consistent across
 * platforms rather than each platform inventing its own scale independently.
 * `ambientColor`/`spotColor` alpha is the source rgba opacity converted to a
 * 0-255 byte (e.g. 0.10 -> 0x1A). `topX` shadows (cast upward, e.g.
 * trays/sheets) reuse the same colors as `bottomX` - Compose has no
 * directional light source to flip for an "upward" shadow. `tilesHover` is
 * the one token with a colored (brand purple) shadow rather than black -
 * kept in the brand hue rather than approximated to grey.
 */
object PUShadows {
    /** bottomA — '0px 2px 2px 0px rgba(0,0,0,0.03)' — subtle card lift */
    fun Modifier.shadowBottomA(shape: Shape) =
        this.shadow(elevation = 1.dp, shape = shape, ambientColor = Color(0x08000000), spotColor = Color(0x08000000))

    /** bottomB — '0px 2px 2px 0px rgba(0,0,0,0.06)' — elevated card */
    fun Modifier.shadowBottomB(shape: Shape) =
        this.shadow(elevation = 1.dp, shape = shape, ambientColor = Color(0x0F000000), spotColor = Color(0x0F000000))

    /** bottomC — '0px 2px 10px 0px rgba(0,0,0,0.05)' — floating card / button */
    fun Modifier.shadowBottomC(shape: Shape) =
        this.shadow(elevation = 3.dp, shape = shape, ambientColor = Color(0x0D000000), spotColor = Color(0x0D000000))

    /** bottomD — '0px 2px 15px 0px rgba(0,0,0,0.10)' — heaviest bottom lift */
    fun Modifier.shadowBottomD(shape: Shape) =
        this.shadow(elevation = 4.dp, shape = shape, ambientColor = Color(0x1A000000), spotColor = Color(0x1A000000))

    /** topA — '0px -2px 2px 0px rgba(0,0,0,0.03)' — tray / nav bottom sheet */
    fun Modifier.shadowTopA(shape: Shape) =
        this.shadow(elevation = 1.dp, shape = shape, ambientColor = Color(0x08000000), spotColor = Color(0x08000000))

    /** topB — '0px -2px 2px 0px rgba(0,0,0,0.06)' */
    fun Modifier.shadowTopB(shape: Shape) =
        this.shadow(elevation = 1.dp, shape = shape, ambientColor = Color(0x0F000000), spotColor = Color(0x0F000000))

    /** topC — '0px -2px 2px 0px rgba(0,0,0,0.10)' */
    fun Modifier.shadowTopC(shape: Shape) =
        this.shadow(elevation = 1.dp, shape = shape, ambientColor = Color(0x1A000000), spotColor = Color(0x1A000000))

    /** topD — '0px -2px 15px 0px rgba(0,0,0,0.10)' — heaviest top lift */
    fun Modifier.shadowTopD(shape: Shape) =
        this.shadow(elevation = 4.dp, shape = shape, ambientColor = Color(0x1A000000), spotColor = Color(0x1A000000))

    /** sideNav — '4px 0px 15px 0px rgba(0,0,0,0.20)' — side navigation panel */
    fun Modifier.shadowSideNav(shape: Shape) =
        this.shadow(elevation = 6.dp, shape = shape, ambientColor = Color(0x33000000), spotColor = Color(0x33000000))

    /** tilesActive — '0px 10px 15px -3px rgba(0,0,0,0.10)' — active/selected tile */
    fun Modifier.shadowTilesActive(shape: Shape) =
        this.shadow(elevation = 4.dp, shape = shape, ambientColor = Color(0x1A000000), spotColor = Color(0x1A000000))

    /** tilesHover — '0px 10px 15px -3px rgba(116,88,253,0.40)' — hovered tile, brand-coloured glow */
    fun Modifier.shadowTilesHover(shape: Shape) =
        this.shadow(elevation = 4.dp, shape = shape, ambientColor = Color(0x667458FD), spotColor = Color(0x667458FD))
}
